### PR TITLE
Mdsolver test exclude

### PR DESCRIPTION
**Pull request by amartyads**
Added checks in the CMakeLists to only compile tests for the currently selected MD solver. This circumvents having many ifdefs in all tests. I've added two sample tests to make sure it works.

If we'd rather have ifdefs, we can discuss it here.

### CODE REVIEW

**Review by Thinkpiet — CHANGES_REQUESTED** [review]
Please improve the regex exclude, this is dangerous. (And difficult to debug, if something unexpected happens...)

Otherwise just small comments, all ok. I like the idea of CMake being responsible for choosing which MD unit tests to compile, instead of having a lot of weird preprocessor macros :-) This is much better.

File: CMakeLists.txt [hunk 1]
```diff
@@ -790,8 +790,20 @@ endif()
 # Testing
 #
 if (BUILD_TESTING)    
-  file(GLOB_RECURSE TEST_SOURCES "test/*Test.cpp")
+  file(GLOB_RECURSE TEST_SOURCES_TOP "test/*Test.cpp")
+  list(FILTER TEST_SOURCES_TOP EXCLUDE REGEX "impl") # remove tests corresponding to specific MD solvers
```
[the remark this answers is not in the document]
Comment (amartyads) [reply CMakeLists.txt:794]: Done, the earlier regex was bad, I agree, I didn't think of those cases

File: test/unit/coupling/interface/impl/SimpleMD/SimpleMDMoleculeTest.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		{
+			for(int j = -4000; j < 4001; j+=100)
+			{
+				for(int k = -4000; k < 4001; k+=100)
+				{
+					tarch::la::Vector<3,double> velocity{i+0.5,j+0.5,k+0.5};
+
+					simpleMolecule.setVelocity(velocity);
+					tarch::la::Vector<3,double> storedVelocity = simpleMolecule.getVelocity();
+					CPPUNIT_ASSERT_MESSAGE( "velocity assersion",storedVelocity == velocity );
```
[the remark this answers is not in the document]
Comment (amartyads) [reply test/unit/coupling/interface/impl/SimpleMD/SimpleMDMoleculeTest.cpp:41]: I think it's okay to make that assumption since `la::Vector`'s overloaded `==` operator has the signature `template <int size, class T> bool operator==(const Vector<size, T>& l, const Vector<size, T>& r)` and internally it's doing the same thing anyway, so the assumption would be that the `==` operator in `la::Vector` is tested. Do you think that's valid? If not, I'll use `CPPUNIT_ASSERT_DOUBLES_EQUAL`, no problem

[the remark this answers is not in the document]
Comment (Thinkpiet) [reply test/unit/coupling/interface/impl/SimpleMD/SimpleMDMoleculeTest.cpp:41]: Maybe my comment was not clear. I try again. Your code was assuming that

1.  `simpleMolecule.setVelocity()` accepts a `Vector<3,double>`. This is fine, since this is `SimpleMDMoleculeTest` and `simpleMolecule` is an `interface::SimpleMDMolecule`. So we definitely want to make assumptions about `simpleMolecule` and test them here.
2. `simpleMolecule.getVelocity()` returns a `Vector<3,double>`. Also fine.
3. `Vector<3,double>::operator==()` works as expected and compares the components using the `==` operator. Also fine, this should be tested for `la::Vector` elsewhere, here you can just assume that it works. 

So all of this is fine. My point was:

- By asserting that `storedVelocity == velocity`, since `simpleMolecule` is a wrapper around `simplemd::Molecule myMolecule` you implicitly also assumed that `myMolecule` internally has something like `Vector<3,double> _velocity` (or e.g. `double _velX, _velY, _velZ`). But `_velocity` is `private`.
- If we would change simplemd/Molecule.h to use e.g. `Vector<3,float> _velocity` or to use fixed-point math internally (keeping the signatures of all public methods constant), then the `SimpleMDMoleculeTest` would fail.
- I do not think that this test should fail in this case, since here you want to test `coupling::interface::SimpleMDMolecule` but not `simplemd::Molecule`, so this test should be independent of the implementation details of `simplemd::Molecule`.

**Review by Thinkpiet — APPROVED** [review]

File: test/unit/coupling/interface/impl/ls1/LS1StaticCommDataTest.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		coupling::interface::LS1StaticCommData::getInstance().setConfigFilename(filename);
+		CPPUNIT_ASSERT( coupling::interface::LS1StaticCommData::getInstance().getConfigFilename() == filename );
+
+		std::array<double,3> offsets({5,5,5});
+		for(int i = 0; i < 3; i++)
+			coupling::interface::LS1StaticCommData::getInstance().setBoxOffsetAtDim(i,offsets[i]);
+		for(int i = 0; i < 3; i++)
+			CPPUNIT_ASSERT( coupling::interface::LS1StaticCommData::getInstance().getBoxOffsetAtDim(i) == offsets[i] );
+	}
+	void testParallelFunctions()
```
Comment (LouieVoit) [inline test/unit/coupling/interface/impl/ls1/LS1StaticCommDataTest.cpp:42]: Wouldn't it be better to have this `#if (COUPLING_MD_PARALLEL == COUPLING_MD_YES)` line 41 and add the test line 18 only if compiling in parallel ? Otherwise we would have an empty test when not parallel that would succeed, could be confusing.

File: test/unit/coupling/interface/impl/SimpleMD/SimpleMDMoleculeTest.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (LouieVoit) [reply test/unit/coupling/interface/impl/SimpleMD/SimpleMDMoleculeTest.cpp:41]: But the `simpleMolecule.getVelocity()` would always return a `Vector<3,double>` even if internally the `simplemd::Molecule` uses float or something else. I don't see the problem here.

[the remark this answers is not in the document]
Comment (amartyads) [reply test/unit/coupling/interface/impl/SimpleMD/SimpleMDMoleculeTest.cpp:41]: I see what Piet means
Let's say that the `simplemd::Molecule` internally converts `double` to `float`, and we lose some precision. Let's say that `getVelocity()` upcasts the `float` to `double` and returns it. If we check for equality at bit level, our tests will fail. But if we check numerical representation with a tolerance of 1e-6, it might still pass.

So the question is, what do we want this test to test? If we want it to enforce that absolutely nothing changes at bitlevel, then my way works. But if we want the `simplemd::Molecule` class to have the flexibility to do whatever it wants, as long as it gives us 6 digits after the decimal, then we do it Piet's way.

I'll change it to be Piet's way, then we can look at it further.

File: test/unit/coupling/interface/impl/ls1/LS1StaticCommDataTest.cpp (hunk 3 above)
Comment (amartyads) [reply test/unit/coupling/interface/impl/ls1/LS1StaticCommDataTest.cpp:42]: You're right, I'll do it

**Review by LouieVoit — APPROVED** [review]
